Add a search box to the Sample Problems page and a script to build a search file

pstaabp opened the pull request:
This PR has two related features
1. A script in `bin/dev_scripts` to parse all of the sample problems and the POD to build up a "database" (JSON file) of non-common words for a given problem/POD page.  The JSON file contains information to be searched on within the browser using the package MiniSearch. 
2. Add search boxes to the sample-problems/POD pages to return pages that match the search criteria.  The searching is done using MiniSearch. 

There is a file `htdocs/DATA/search.json` that is provided that will be used while searching.  This is not complete but currently should give some sense about how this works.   The script will rebuild this file as needed. 

Note: a subsequent PR will be posted to update the POD in the macros to give the search tools needed terms.

pstaabp committed:
Initial Checkin of search code

This PR has two related features
1. A script in `dev_scripts` to parse all of the sample problems and the POD to build up a "database" of non-common words for a given problem/POD page.
2. Add search boxes to the sample-problems/POD pages to return pages that match the search criteria.

pstaabp committed:
run prettier-format

pstaabp committed:
Parse the POD from macro files .

pstaabp committed:
improve the search and add a "no results"

Also add ww-update script to call OPL-update and build-search-json.pl

dlglin commented:
When I start typing I get suggestions, but if I type more after that then the suggestions disappear.  Then if I type even more they reappear.

pstaabp commented:
Added an `update-ww.pl` script that will 
1. run `OPL-update` without the interactive loadOPLstats part of the script
2. run `build-search-json.pl` to generate the search-json for the POD/sample probs. 

Note: this will need to be added to the upgrade instructions. 

So now to test, run `update-ww.pl`.  To skip the OPL-update use the `-s` flag.  Then test the search box on the sample problems page.

drgrice1 commented on bin/build-search-json.pl:
```diff
@@ -0,0 +1,262 @@
+die "The build options must be one of (all, macros, samples). The value $build is not valid."
+	if ((grep { $_ eq $build } qw/all macros samples/) != 1);
+
+my $ww_root = $ENV{WW_ROOT};
+$ww_root = Mojo::File->curfile->dirname->dirname->realpath unless defined($ww_root);
```
Drop the usage of the `WW_ROOT` environment variable entirely.  Just use the correct path relative to the location of this file.  The correct environment variable that we have always used (but should not be used in general) is `WEBWORK_ROOT`, so definitely do not introduce a new one that no one will have set.  Also, there is no need for the `realpath` call on the next line.  So this should be
```suggestion
my $ww_root = Mojo::File->curfile->dirname->dirname;
```

drgrice1 commented on bin/build-search-json.pl:
```diff
@@ -0,0 +1,262 @@
+use Pod::Simple::SimpleTree;
+
+my $build   = "all";
+my $pg_root = $ENV{PG_ROOT};
```
Also drop the usage of the `PG_ROOT` environment variable.  Once the webwork2 root directory is determined relative to this script, then load the `webwork2.mojolicious.yml` file and get the PG root directory from the settings.  See how this is done in the `bin/check_latex` script for example (or in the `bin/webwork2` app script itself).

drgrice1 commented on bin/download-OPL-metadata-release.pl:
```diff
@@ -80,8 +80,12 @@ BEGIN
 die "The directory $libraryDirectory does not exist or is not writable."
 	if (!-d $libraryDirectory || !-w $libraryDirectory);
 
+warn $ENV{OPL_GIT_REPOSITORY};
```
Obviously this line needs to be deleted.

drgrice1 commented on bin/update-ww.pl:
```diff
@@ -0,0 +1,72 @@
+
+=head1 NAME
+
+update-ww.pl -- run all needed scripts when webwork is updated.
```
Don't add this script.  The `OPL-update` script is already a script that just runs other scripts.  Now you are going to add another one that runs that script and one more?  Find another way to do it.

In addition the options for this script are not well explained, and there is no `-h` help flag.

drgrice1 replied:
Yes, that is what I actually meant with the suggestion.  That the `OPL-update` includes running this script.  Not the creation of a new script.  It is a bit of a stretch that this goes under the description of "updating the OPL", but the point is that everything be handled by a single script.  However, that is minor, and the point is to not need to change the instructions for installation/upgrade.

Although, I have also though about just removing the need for this new script entirely.  Have you actually benchmarked the script?  How bad would performance be if the files were dynamically combed at runtime, and the search actually done live?  If it isn't bad, or with some tweaking could be done efficiently at runtime, then that is the ideal way to go.

drgrice1 replied:
Have you considered that the reason the script takes 5-10 seconds to run is that your code is highly inefficient?  And that improvements could be made to make it run in a fraction of the time?  I know this to be true in fact, and I made a small change that makes it so that running the script takes around 4 tenths of a second.  I am certain better can be done with more effort.

Have you also actually looked at the result of running the script and the words that are being listed in the JSON file?  That clearly needs more thought.  I am pretty sure that we don't want the word "`exppi" which is listed (for example from the ProvingTrigIdentities.pg sample problem) because "`exp(pi" appears in a file and the script removes the parenthesis.

drgrice1 commented on bin/update-ww.pl:
```diff
@@ -0,0 +1,72 @@
+use Mojo::File;
+use Getopt::Long qw(:config bundling);
+
+# BEGIN {
+# 	use Env        qw(WEBWORK_ROOT);
+# }
```
When lines like this are in new files in a pull request, that means this pull request is really not ready for review.  Look over the code, and clean it up.

drgrice1 reviewed (changes requested):
It seems that you need to run `npm install minisearch`, and then add the changed `package.json` and `package-lock.json` files to this pull request.

I haven't delved into this in much depth yet.  I can see that this pull request clearly needs a lot of work before it is really ready for review.

pstaabp committed:
Remove the update-ww script, return the download-OPL script ...

to it's previous state and now run the parse-search-json.pl script upon startup of webwork2.

Restore the minisearch package.

pstaabp commented:
If you put in a PR to this branch, I'll deal with the conflicts.  I can made it run in the webwork2 script and probably is unnecessary and then can remove that.

drgrice1 commented:
Most likely I am just going to take this over and put in a completely new pull request.  I am sorry, but there is a lot wrong with this pull request, and almost everything needs to be completely rewritten.  There aren't going to be any scripts.  Everything is going to be done efficiently and dynamically directly from the code.

pstaabp commented:
That's fine @drgrice1
